### PR TITLE
Improve emq_coap code

### PR DESCRIPTION
**Pull request by grutabow**
1) add unit test cases.
    2) fix a coap encoding bug which generate option list in wrong order.
    3) make parser more stable to prevent format error.
    4) introduce percent-encoding to topic and message payload, which make topic suitable for special character, such as '/'
    5) fix a crash during unsubscribing.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -1,8 +1,8 @@
 PROJECT = emq_coap
 PROJECT_DESCRIPTION = CoAP Gateway
-PROJECT_VERSION = 0.2.1
+PROJECT_VERSION = 0.2.2
```
Comment (emqplus) [inline Makefile]: Change the version to 0.2.7

File: README.md [hunk 2]
```diff
@@ -12,16 +12,15 @@ File: etc/emq_coap.conf
 ```
 coap.server = 5683
 
-coap.prefix.mqtt = mqtt
-
-coap.handler.mqtt = emq_coap_gateway
+coap.prefix.set1 = mqtt
+coap.handler.set1 = emq_coap_gateway
```
Comment (emqplus) [inline README.md]: Should rename the 'set1' to 1:
```
coap.prefix.1 = mqtt
coap.handler.1 = emq_coap_gateway
```

File: etc/emq_coap.conf [hunk 3]
```diff
@@ -4,7 +4,13 @@
 
 coap.server = 5683
 
-coap.prefix.mqtt = mqtt
+coap.prefix.set1 = mqtt
+coap.handler.set1 = emq_coap_gateway
```
Comment (emqplus) [inline etc/emq_coap.conf]: Remove the `set` prefix.

File: src/emq_coap.app.src [hunk 4]
```diff
@@ -0,0 +1,9 @@
+{application,emq_coap,
+             [{id,"emq_coap"},
+              {vsn,"0.2.2"},
```
Comment (emqplus) [inline src/emq_coap.app.src]: Version 0.2.7

File: src/emq_coap.app.src [hunk 5]
```diff
@@ -0,0 +1,9 @@
...
+              {description,"EMQTT coap gateway"},
```
Comment (emqplus) [inline src/emq_coap.app.src]: EMQ CoAP Gateway

File: src/emq_coap_broker_api.erl [hunk 6]
```diff
@@ -0,0 +1,58 @@
+%%--------------------------------------------------------------------
+%% Copyright (c) 2016 Feng Lee <feng@emqtt.io>. All Rights Reserved.
```
Comment (emqplus) [inline src/emq_coap_broker_api.erl]: Copyright (c) 2016-2017

File: README.md (hunk 2 above)
Comment (grutabow) [inline README.md]: Done.

**Pull request merged**